### ISSUE TITLE
The sorting on the "Date" column 

### ISSUE THREAD

**Issue by riik91760** [moneymanagerex/moneymanagerex#5800]
> Hello. The sorting on the "Date" column does not take into account the previous sorting. The sorting on the "Date" column is always associated with a sorting on the "Operation number" column. Kind regards. Eric

![screen_20230407_112656](https://user-images.githubusercontent.com/130142043/230592223-5d0cd739-278c-4730-8393-65b1e393c68e.jpg)

![screen_20230407_115052](https://user-images.githubusercontent.com/130142043/230592458-1f10fd58-c657-4d73-bddd-86350353e70d.jpg)

![screen_20230407_115908](https://user-images.githubusercontent.com/130142043/230592515-72aaf2b8-acbf-4679-b382-986648ac9294.jpg)

![screen_20230407_120204](https://user-images.githubusercontent.com/130142043/230592580-44470d50-bad1-4934-9236-0ff63975c4cd.jpg)

![screen_20230407_120327](https://user-images.githubusercontent.com/130142043/230592654-a50c9389-5ae2-404c-a128-21cf2b8be70b.jpg)

![screen_20230407_120455](https://user-images.githubusercontent.com/130142043/230592708-b08a597a-750c-47d7-b892-74151a3feae2.jpg)

![screen_20230407_120608](https://user-images.githubusercontent.com/130142043/230592755-0dd57e6c-eaec-4a1e-a8ab-411a61e94557.jpg)

> Veuillez ne pas effacer les informations jointes sous ce texte.

<hr> • Version: 1.6.3 <br>
 • Built: Janv. 31 2023 12:20:31<br>
 • db 17 (aes128cbc)<br>
 • git 48d8562 (2023-01-31)<br>
<br>
Libs:<br>
 • wxWidgets 3.2.1 (wxMSW 10.0)<br>
 • wxSQLite3 4.9.1 (SQLite 3.40.0)<br>
 • RapidJSON 1.1.0<br>
 • Lua 5.3.5<br>
 • libcurl/7.82.0-DEV Schannel<br>
 • gettext 0.19.8.1<br>
 • apexcharts.js<br>
<br>
Built with:<br>
 • CMake 3.16.2<br>
 • MSBuild 15.9.21.664<br>
 • MSVS15.0<br>
 • MSVSC  19.16.27035.0<br>
 • Windows SDK 10.0.14393.0<br>
<br>
OS:<br>
 • Windows 10 (build 19044), 64-bit edition<br>
 •  French (France) (windows-1252)<br>
 • 1920x1080 32bit 96x96ppi

**Comment by tactilis** [comment]
You are correct but this is by design.

When Secondary Sort (i.e. the ability to have a Primary and a Secondary sort column) was first implemented, we received a large number of 'bug' reports saying that MMEX was calculating the Balance incorrectly even though people were sorting by Date.

This was because they had Date as the primary sort column and a column _other than ID_ (No. opé) as the secondary sort.

The Balance column will only be correct if the sorting is Date / ID (both ascending or both descending).

A change was implemented in the next release to force the sort order to be Date / ID when clicking on the Date header. The 'bug' reports then stopped.

**Comment by riik91760** [comment]
I understand. However I think that forcing the secondary sort "ID" with "Date" is not the best solution. This solution makes the secondary sort column lose a lot of interest. Perhaps it would have been simpler just to tell the applicants that to get a correct "Balance" column they should sort on "ID" and then "Date". In my usage, I use the "Number" column to indicate the time of certain transactions and therefore I would like to sort on "Date" + "Number". In this case, I don't care about the accuracy of the information in the "Balance" column. The debate is open.

Translated with www.DeepL.com/Translator (free version)

**Comment by tactilis** [comment]
> Perhaps it would have been simpler just to tell the applicants that to get a correct "Balance" column they should sort on "ID" and then "Date".

We did, of course, do that. Repeatedly.
But still the bug reports came complaining of an incorrect figures in the Balance column.

<br>

> The debate is open.

Listening.

Could you explain your use case in more detail please?

What is the significance of the transaction time for you?

Why does the Balance not matter to you?

Could you show a screenshot with a few example transactions to illustrate your usage please?

Could you perhaps enter the transactions in time order, in which case Date + ID sorting would be correct?

**Comment by n-stein** [comment]
I've made this suggestion before (https://github.com/moneymanagerex/moneymanagerex/issues/5122#issuecomment-1256408856_), but my preference is to sort by the user's secondary sort column then by date for the balance calculation rather than the constant ID/Date sort we are forcing. That way we retain the usefulness of the secondary sort and the balance is still correct for the top row of each date.

**Comment by riik91760** [comment]
There is something I don't understand; The amount in the "Balance" column is calculated when the transaction is entered. Whatever sorting is requested, the amount displayed in the "Balance" column of a given transaction remains the same.
As for my use, when I enter a credit card bill, I enter in "Number" the time of the transaction and the last 4 digits of the card.
Entering the transactions in chronological order is not a solution. It often happens that I enter my bills in one session, then my wife's bills in another session, then her cheques several days or weeks later. There is also always the possibility of forgetting to enter transactions late.

Translated with www.DeepL.com/Translator (free version)

![screen_20230407_142231](https://user-images.githubusercontent.com/130142043/230608630-eb260fb0-96c4-4c38-b762-21e30ffad583.jpg)

**Comment by tactilis** [comment]
> I've made this suggestion before (https://github.com/moneymanagerex/moneymanagerex/issues/5122#issuecomment-1256408856), but my preference is to sort by the user's secondary sort column then by date for the balance calculation rather than the constant ID/Date sort we are forcing. That way we retain the usefulness of the secondary sort and the balance is still correct for the top row of each date.

Yes... but note the comments that followed yours.

We got such a backlash when users were finding that the MMEX running balance did not match their bank account running balance due to a confusing sort order that the majority didn't understand. Like @whalley (in the comments), I would be very wary about changing the calculation.

**Comment by riik91760** [comment]
For me the only balance values that matter are the ones indicated at the top :
- "Reconciliation balance" which should correspond to the balance shown at the end of my last bank statement after reconciliations.
- "Account balance" which corresponds to the balance I have at the current state of the entries.

**Comment by riik91760** [comment]
The time of the transaction associated with the last 4 digits of a card constitutes an identifier which, when a transaction is duplicated, allows me to match them. It is important for certain transactions, notably for fuel purchases in connection with trip tracking applications and their costs.
Yes, it would be useful to be able to specify the time of the transactions when entering them.
In my capture, the transactions are recent or upcoming.
I only reconcile once a month from the bank statements.
I don't import my transactions directly from my banks because their information is too basic and not categorized.

Translated with www.DeepL.com/Translator (free version)

**Comment by tactilis** [comment]
@riik91760
> The time of the transaction associated with the last 4 digits of a card constitutes an identifier which, when a transaction is duplicated, allows me to match them. It is important for certain transactions, notably for fuel purchases in connection with trip tracking applications and their costs.
Yes, it would be useful to be able to specify the time of the transactions when entering them.

It sounds as though you might benefit from defining some custom fields (see Tools | Custom Fields Manager) to record card number and time, instead of merging them into the Number field.

Custom field definition:

![image](https://user-images.githubusercontent.com/4060092/230624213-ee16923a-10e6-4b92-9f00-22c79f371a00.png)

<br>

Field usage:
 
![image](https://user-images.githubusercontent.com/4060092/230624272-1fa2068d-e69e-40bc-8b4c-bbcec8a1cb40.png)

You could even define a custom field that was used for trip tracking. Maybe a String field into which you enter the name of the trip.

Just an idea to allow you to make use of the flexibility of MMEX.

If you do use custom fields and come across any aspects that could be improved, please do raise new issues so they can be discussed.

**Comment by riik91760** [comment]
Thank you for this last answer.
This is one aspect and capability of MMEX that I had not yet used. I will look into these options very seriously.
I wonder if I should also create a custom "Transaction Date" field so that I can do custom sorting on this new field and the new "Transaction Time" field.
Can custom fields be exported and then imported after update outside the application?
I'll come back to you if I don't find my happiness in the application help.

Translated with www.DeepL.com/Translator (free version)

**Comment by tactilis** [comment]
> I wonder if I should also create a custom "Transaction Date" field so that I can do custom sorting on this new field and the new "Transaction Time" field.

That would work. So if it's important to you, then yes you could do it.

> Can custom fields be exported and then imported after update outside the application?

Not yet. Although the rate at which @n-stein is working on improving import (for example, adding automatic Payee and Category matching which will be released soon), then import and export of custom fields _might_ come along in the fairly near future.

There is already an issue to track it: https://github.com/moneymanagerex/moneymanagerex/issues/5401
It just needs time and motivation. Your request helps.

> I'll come back to you if I don't find my happiness in the application help.

Note that the MMEX Help is out of date. It's not wrong but it lacks descriptions of the newer features.
